front/src/modules/trainschedule/components/TimetableV2/hooks.ts

```diff
@@ -0,0 +1,144 @@
+  debouncedFilter: string,
+  debouncedRollingstockFilter: string,
+  validityFilter: string,
+  selectedTags: Set<string | null>
+) => {
+  const infraId = useInfraID();
+
+  const { data: trainSchedulesSummary } =
+    enhancedEditoastApi.endpoints.getV2TrainScheduleSimulationSummary.useQuery(
+      {
+        infra: infraId as number,
+        ids: trainIds,
+      },
+      { skip: !infraId || !trainIds.length }
+    );
+
+  const { data: trainSchedules } = enhancedEditoastApi.endpoints.getV2TrainSchedule.useQuery(
+    {
+      ids: trainIds,
+    },
+    {
+      skip: !trainIds.length,
+    }
+  );
+
+  // We fetch all RS to get the data we need for the advanced filters
+  const { data: { results: rollingStocks } = { results: [] } } =
+    enhancedEditoastApi.endpoints.getLightRollingStock.useQuery({ pageSize: 1000 });
+
+  useEffect(() => {
+    if (trainSchedulesSummary && trainSchedules) {
+      const formatedTrains: TrainScheduleWithDetails[] = trainSchedules
+        .filter((trainSchedule) => {
+          if (!keepTrain(trainSchedule, debouncedFilter)) return false;
+
+          const trainSummary = trainSchedulesSummary[trainSchedule.id];
+
+          // Apply validity filter
+          if (validityFilter === 'valid' && trainSummary.status !== 'success') return false;
+          if (validityFilter === 'invalid' && trainSummary.status === 'success') return false;
+
+          // Apply tag filter
+          if (
+            trainSchedule.speed_limit_tag &&
+            selectedTags.size > 0 &&
+            !selectedTags.has(extractTagCode(trainSchedule.speed_limit_tag))
+          )
+            return false;
+
+          // Apply rolling stock filter
+          const rollingStock = rollingStocks.find(
+            (rs) => rs.name === trainSchedule.rolling_stock_name
+          );
+          const {
+            detail = '',
+            family = '',
+            reference = '',
+            series = '',
+            subseries = '',
+          } = rollingStock?.metadata || {};
+          if (
+            ![detail, family, reference, series, subseries].some((v) =>
+              v.toLowerCase().includes(debouncedRollingstockFilter.toLowerCase())
+            )
+          )
+            return false;
+
+          return true;
+        })
+        .sort(
+          (trainA, trainB) =>
+            new Date(trainA.start_time).valueOf() - new Date(trainB.start_time).valueOf()
+        )
+        .map((trainSchedule) => {
+          const rollingStock = rollingStocks.find(
+            (rs) => rs.name === trainSchedule.rolling_stock_name
+          );
+          const trainSummary = trainSchedulesSummary[trainSchedule.id];
+          const formattedStartTimeMs = isoDateToMs(trainSchedule.start_time);
+          const arrivalTime =
+            trainSummary.status === 'success'
+              ? msToIsoDate(formattedStartTimeMs + trainSummary.time, true)
+              : '';
```

Comment:
Create a utils object like this: 
```suggestion
          const otherProps =
            trainSummary.status === 'success'
              ? {
                  arrivalTime: msToIsoDate(formattedStartTimeMs + trainSummary.time, true),
                  duration: trainSummary.time,
                  pathLength: formatKmValue(trainSummary.length, 'millimeters', 1),
                  mechanicalEnergyConsumed: jouleToKwh(trainSummary.energy_consumption, true),
                }
              : {
                  arrivalTime: '',
                  duration: 0,
                  pathLength: '',
                  mechanicalEnergyConsumed: 0,
                  invalidReason: trainSummary.status,
                };
```
Then you can just destructure this object:
```
        const formatedTrain: TrainScheduleWithDetails = {
            ...,
            ...otherProps,
          };
```

Reply:
Done